### PR TITLE
subAtlas(Polygon) missing reverse edge and performance

### PR DESCRIPTION
**Pull request by matthieun**
### Description:

This fixes two issues:
- `subAtlas(Polygon)` would in some cases include a reverse edge without its master edge, in a very corner case where JTS would match the reverse edge to the `Polygon`, but not the master edge.
```
org.openstreetmap.atlas.geography.atlas.exception.AtlasIntegrityException: Cannot build an Atlas with a negative edge without its positive counterpart: -XXXXXXXXX
	at org.openstreetmap.atlas.geography.atlas.packed.PackedAtlasBuilder.lambda$verifyNegativeEdgesHaveMasterEdge$1(PackedAtlasBuilder.java:319)
	at java.lang.Iterable.forEach(Iterable.java:75)
	at org.openstreetmap.atlas.geography.atlas.packed.PackedAtlasBuilder.verifyNegativeEdgesHaveMasterEdge(PackedAtlasBuilder.java:312)
	at org.openstreetmap.atlas.geography.atlas.packed.PackedAtlasBuilder.get(PackedAtlasBuilder.java:211)
	at org.openstreetmap.atlas.geography.atlas.BareAtlas.subAtlas(BareAtlas.java:629)
```
- Improve `subAtlas` performance by caching the edges intersecting the polygon, instead of re-computing them 3 times.

### Potential Impact:

A more accurate `subAtlas` in very corner cases. A faster `subAtlas` in most cases.

### Unit Test Approach:

Added one unit test that exposes the corner case, and verifies that the master edge gets added.

### Test Results:

Tests succeed.

------

In doubt: [Contributing Guidelines](CONTRIBUTING.md)

### CODE REVIEW

File: src/main/java/org/openstreetmap/atlas/geography/atlas/BareAtlas.java [hunk 1]
```diff
@@ -914,6 +942,34 @@ private void checkRelationMembersAndIndexRelation(final Relation relation,
...
+                memberIdentifiersIntersecting.add(intersecting);
+                return Iterables.stream(source).map(member ->
+                {
+                    intersecting.add(member.getIdentifier());
+                    return member;
+                }).collect();
+            }
+            else
+            {
+                return (Iterable<Member>) Iterables.stream(memberIdentifiersIntersecting.get(0))
```
Comment (adahn6) [inline src/main/java/org/openstreetmap/atlas/geography/atlas/BareAtlas.java]: I'm confused about the purpose of the List wrapper around the Set here. Since this pattern limits the List to fetching only one set, why not just use the Set directly?

File: src/main/java/org/openstreetmap/atlas/geography/atlas/BareAtlas.java [hunk 2]
```diff
@@ -914,6 +942,34 @@ private void checkRelationMembersAndIndexRelation(final Relation relation,
...
+        // A supplier that will effectively cache all the members intersecting that polygon. This is
+        // thread safe because the cache is internal to the supplier's scope.
+        @SuppressWarnings("unchecked")
+        final Supplier<Iterable<Member>> result = () ->
+        {
+            if (memberIdentifiersIntersecting.isEmpty())
+            {
+                final Set<Long> intersecting = new HashSet<>();
+                memberIdentifiersIntersecting.add(intersecting);
+                return Iterables.stream(source).map(member ->
```
Comment (lucaspcram) [inline src/main/java/org/openstreetmap/atlas/geography/atlas/BareAtlas.java]: I find it a little confusing that the call to `map` does not actually perform a mapping operation, and instead produces a side effect on some external global state. I think you are doing this to avoid calling `Iterables.stream()` twice (Once for a `forEach` and once for the `collect`). So here, you are using `map` as a modified `forEach` that returns an identical stream (so that `collect` can be called in the same line)?

If this is the case, maybe leave a comment that explains this, since this use of `map` is bit non-idiomatic :)

File: src/main/java/org/openstreetmap/atlas/geography/atlas/BareAtlas.java [hunk 3]
```diff
@@ -914,6 +942,34 @@ private void checkRelationMembersAndIndexRelation(final Relation relation,
         }
     }
 
+    private <Member extends AtlasEntity> Supplier<Iterable<Member>> getCachingSupplier(
+            final Iterable<Member> source, final ItemType type)
+    {
+        final List<Set<Long>> memberIdentifiersIntersecting = new ArrayList<>();
```
Comment (lucaspcram) [inline src/main/java/org/openstreetmap/atlas/geography/atlas/BareAtlas.java]: Could this also work as an `Optional`? Might make things a little clearer, since it's a singleton `List` anyway.

File: src/main/java/org/openstreetmap/atlas/geography/atlas/BareAtlas.java (hunk 2 above)
Comment (adahn6) [reply src/main/java/org/openstreetmap/atlas/geography/atlas/BareAtlas.java]: +1 to this

File: src/main/java/org/openstreetmap/atlas/geography/atlas/BareAtlas.java [hunk 4]
```diff
@@ -560,9 +568,24 @@ public SnappedEdge snapped(final Location point, final Distance threshold)
...
 
-        // Add the edges
-        edgesIntersecting(boundary).forEach(
-                edge -> builder.addEdge(edge.getIdentifier(), edge.asPolyLine(), edge.getTags()));
+        // Add the edges. Use a consumer that makes sure master edges are always added first.
+        final Consumer<Edge> edgeAdder = edge ->
+        {
+            if (builder.peek().edge(edge.getIdentifier()) == null)
+            {
+                if (edge.getIdentifier() != 0 && edge.hasReverseEdge())
```
[the remark this answers is not in the document]
Comment (lucaspcram) [reply src/main/java/org/openstreetmap/atlas/geography/atlas/BareAtlas.java]: I agree.

It seems to me that swapping `edge.getIdentifier() != 0` with `!isMasterEdge()` would correct this code segment.

File: src/main/java/org/openstreetmap/atlas/geography/atlas/BareAtlas.java (hunk 1 above)
Comment (matthieun) [reply src/main/java/org/openstreetmap/atlas/geography/atlas/BareAtlas.java]: Because the object that goes from outside to inside the lambda has to be final. I cannot make the set final as I use the fact it might not be there as a cause for initializing it. So wrapping it in a final list is a trick that might be ugly but works :) I am open to any other more elegant suggestion!

File: src/main/java/org/openstreetmap/atlas/geography/atlas/BareAtlas.java [hunk 5]
```diff
@@ -560,9 +568,24 @@ public SnappedEdge snapped(final Location point, final Distance threshold)
         nodesWithin(boundary, hasNode.negate()).forEach(
                 node -> builder.addNode(node.getIdentifier(), node.getLocation(), node.getTags()));
 
-        // Add the edges
-        edgesIntersecting(boundary).forEach(
-                edge -> builder.addEdge(edge.getIdentifier(), edge.asPolyLine(), edge.getTags()));
+        // Add the edges. Use a consumer that makes sure master edges are always added first.
```
[the remark this answers is not in the document]
Comment (matthieun) [reply src/main/java/org/openstreetmap/atlas/geography/atlas/BareAtlas.java]: So if the `1` and `-1` have not been added yet there are multiple use cases:
- The consumer is told to add `1` first.
  - If `1` has a reverse edge, it will add `-1` (if not added already), then `1` (if not added already)
  - If `1` does not have a reverse edge, it will just add `1` (if not added already).
- The consumer is told to add `-1` first.
  - It will add `1` first (if not added already) then `-1` (if not added already)

File: src/main/java/org/openstreetmap/atlas/geography/atlas/BareAtlas.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (matthieun) [reply src/main/java/org/openstreetmap/atlas/geography/atlas/BareAtlas.java]: We have a lot of unit tests that unfortunately create edges with id 0 using `@TestAtlas`. The whole idea of having a `+` and `-` for forward and reverse edges breaks down unfortunately for edge 0, and all the tests break. This might be worth creating an issue to track it, however the impact is pretty small. Thoughts?

File: src/main/java/org/openstreetmap/atlas/geography/atlas/BareAtlas.java (hunk 3 above)
Comment (matthieun) [reply src/main/java/org/openstreetmap/atlas/geography/atlas/BareAtlas.java]: Well it would not work as an optional given the fact that you cannot swap the internals of an Optional object without changing the optional handle... or can you?
Basically, like in https://github.com/osmlab/atlas/pull/251/files#r227476980, the issue is that I can share objects within the lambda only when they are final. Open to other ideas though!

File: src/main/java/org/openstreetmap/atlas/geography/atlas/BareAtlas.java (hunk 2 above)
Comment (matthieun) [reply src/main/java/org/openstreetmap/atlas/geography/atlas/BareAtlas.java]: That works :) I will add a comment.

File: src/main/java/org/openstreetmap/atlas/geography/atlas/BareAtlas.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (MikeGost) [reply src/main/java/org/openstreetmap/atlas/geography/atlas/BareAtlas.java]: That makes sense, it's impossible to know that without you explaining - I was trying to wrap my head around why 0 was there - it's worth putting that as a comment here. Otherwise, is it possible to clean up tests now? If not, creating an issue like you said to clean up the test and update this piece of code to not rely on 0 might be worthwhile.

File: src/main/java/org/openstreetmap/atlas/geography/atlas/BareAtlas.java [hunk 6]
```diff
@@ -514,18 +516,24 @@ public SnappedEdge snapped(final Location point, final Distance threshold)
...
         // buffer on top of the nodes inside the polygon. This mostly avoids resizing.
-        final double nodeRatioBuffer = 1.2;
-        final long nodeNumber = Math.round(Iterables.size(nodesWithin(boundary)) * nodeRatioBuffer);
-        final long edgeNumber = Iterables.size(edgesIntersecting(boundary));
-        final long areaNumber = Iterables.size(areasIntersecting(boundary));
-        final long lineNumber = Iterables.size(linesIntersecting(boundary));
-        final long pointNumber = Iterables.size(pointsWithin(boundary));
-        final long relationNumber = Iterables.size(relationsWithEntitiesIntersecting(boundary));
+        final double ratioBuffer = 1.2;
+        final long nodeNumber = Math.round(Iterables.size(nodesWithin(boundary)) * ratioBuffer);
```
Comment (matthieun) [inline src/main/java/org/openstreetmap/atlas/geography/atlas/BareAtlas.java]: Maybe I should use the caching for all the other types as well!

File: src/main/java/org/openstreetmap/atlas/geography/atlas/BareAtlas.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (lucaspcram) [reply src/main/java/org/openstreetmap/atlas/geography/atlas/BareAtlas.java]: So in sub-bullet point 1 under bullet point 1, you still add the reverse edge before the master edge. Is this OK in this case? I guess I am just confused by the discrepancy between the comment and what actually happens.

Also, I am curious why we need to make sure the master gets added first. Sorry to be probing, I just want to make sure I understand!

File: src/main/java/org/openstreetmap/atlas/geography/atlas/BareAtlas.java (hunk 3 above)
Comment (lucaspcram) [reply src/main/java/org/openstreetmap/atlas/geography/atlas/BareAtlas.java]: Yeah you can't, so that wouldn't work.... Darn

File: src/main/java/org/openstreetmap/atlas/geography/atlas/BareAtlas.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (matthieun) [reply src/main/java/org/openstreetmap/atlas/geography/atlas/BareAtlas.java]: Cleaning up the tests will be a full task. Created an issue #252, will add comments too.

File: src/main/java/org/openstreetmap/atlas/geography/atlas/BareAtlas.java (hunk 1 above)
Comment (adahn6) [reply src/main/java/org/openstreetmap/atlas/geography/atlas/BareAtlas.java]: hmm... wouldn't it be possible to just use the `isEmpty()` check for initializing? a final `Set` can still be modified via `addAll()`, so you could do something like
```
final Set<Long> intersecting = new HashSet<>();
@SuppressWarnings("unchecked")
final Supplier<Iterable<Member>> result = () -> {
     if(intersecting.isEmpty()) {
           intersecting.addAll(Iterables.asSet(source));
     }
     return (Iterable<Member>) Iterables.stream(intersecting);
}
```
Not sure that code compiles or works, just spitballing here

Comment (lucaspcram) [reply src/main/java/org/openstreetmap/atlas/geography/atlas/BareAtlas.java]: I am also curious to know if what @adahn6 suggests works or not.

File: src/main/java/org/openstreetmap/atlas/geography/atlas/BareAtlas.java [hunk 7]
```diff
@@ -514,18 +516,30 @@ public SnappedEdge snapped(final Location point, final Distance threshold)
     {
         final Time begin = Time.now();
 
+        final Supplier<Iterable<Node>> nodesWithin = getCachingSupplier(nodesWithin(boundary),
+                ItemType.NODE);
+        final Supplier<Iterable<Edge>> edgesIntersecting = getCachingSupplier(
+                edgesIntersecting(boundary), ItemType.EDGE);
+        final Supplier<Iterable<Area>> areasIntersecting = getCachingSupplier(
+                areasIntersecting(boundary), ItemType.AREA);
+        final Supplier<Iterable<Line>> linesIntersecting = getCachingSupplier(
```
Comment (MikeGost) [inline src/main/java/org/openstreetmap/atlas/geography/atlas/BareAtlas.java]: Wondering if we can add the same caching strategy to the Predicate subAtlas call.

Comment (matthieun) [reply src/main/java/org/openstreetmap/atlas/geography/atlas/BareAtlas.java]: If the filtering is expensive and we re-filter many times I'd say yes. But we'd want to run the profiler on a big atlas first to make sure it is a worthwhile change.

File: src/main/java/org/openstreetmap/atlas/geography/atlas/BareAtlas.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (lucaspcram) [reply src/main/java/org/openstreetmap/atlas/geography/atlas/BareAtlas.java]: Oh now I understand. The problem was that the `Iterable` returned by `edgesIntersecting` is sometimes missing the master edge. So it doesn't actually matter which is added first, just that we always add the reverse edge/master edge together if both exist.
